compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/calls/Arguments.kt

```diff
@@ -239,27 +242,63 @@ fun Candidate.resolvePlainArgumentType(
             capturedType
 
     // If the argument is of functional type and the expected type is a suspend function type, we need to do "suspend conversion."
-    // TODO: should refer to LanguageVersionSettings.SuspendConversion
-    if (expectedType?.isSuspendFunctionType(session) == true &&
-        argumentTypeForApplicabilityCheck.isBuiltinFunctionalType(session) &&
-        !argumentTypeForApplicabilityCheck.isSuspendFunctionType(session)
-    ) {
-        val typeParameters = argumentTypeForApplicabilityCheck.typeArguments.map { it as ConeKotlinType }
-        argumentTypeForApplicabilityCheck =
-            createFunctionalType(
-                typeParameters.dropLast(1), null, typeParameters.last(),
-                isSuspend = true,
-                isKFunctionType = argumentTypeForApplicabilityCheck.isKFunctionType(session)
-            )
-        substitutor.substituteOrSelf(argumentTypeForApplicabilityCheck)
-        usesSuspendConversion = true
+    if (expectedType != null) {
+        argumentTypeWithSuspendConversion(session, expectedType, argumentTypeForApplicabilityCheck)?.let {
+            argumentTypeForApplicabilityCheck = it
+            substitutor.substituteOrSelf(argumentTypeForApplicabilityCheck)
+            usesSuspendConversion = true
+        }
     }
 
     checkApplicabilityForArgumentType(
         csBuilder, argumentTypeForApplicabilityCheck, expectedType, position, isReceiver, isDispatch, sink, context
     )
 }
 
+private fun argumentTypeWithSuspendConversion(
+    session: FirSession, expectedType: ConeKotlinType, argumentType: ConeKotlinType
+): ConeKotlinType? {
+    // TODO: should refer to LanguageVersionSettings.SuspendConversion
+    // Expect the expected type to be a suspend functional type.
+    if (!expectedType.isSuspendFunctionType(session)) {
+        return null
+    }
+
+    // Either the argument type is a non-suspend functional type
+    if (argumentType.isBuiltinFunctionalType(session) && !argumentType.isSuspendFunctionType(session)) {
+        val typeParameters = argumentType.typeArguments.map { it as ConeKotlinType }
+        return createFunctionalType(
+            typeParameters.dropLast(1), null, typeParameters.last(),
+            isSuspend = true,
+            isKFunctionType = argumentType.isKFunctionType(session)
+        )
+    }
+
+    // Or is a class-like type whose corresponding class is a user-defined and has an overridden `invoke`.
+    if (argumentType is ConeClassLikeType) {
```

Comment:
This part looks very similar to the second part in `findInvokeSymbol`.
Please, try extracting such things into utility methods for better reusing.

Reply:
Utilized at the 2nd commit: https://github.com/JetBrains/kotlin/pull/3718/commits/ebe92c7a6457d21156dca30916b96d1dd6e4e066